
---
title: "Lecture: Exploratory analysis of microbiome data in R/Bioconductor"
author: "Levi Waldron"
date: "June 30, 2017"
output:
  slidy_presentation: default
---

```{r setup, cache=TRUE, echo=FALSE}
library(knitr)
# global chunk options
opts_chunk$set(cache=TRUE, autodep=TRUE)
```

## Outline

- Statistical properties of metagenomic data
- Distances for high dimensional data
- Principal Components and Principal Coordinates Analysis
- Alpha diversity

## Properties of processed microbiome data

- taxonomic data for kingdom, phylum or division, class, order, family, genus, (species)
    + hundreds of rows
- inferred metabolic function
    + hundreds of rows
- calculated alpha diversity

## Properties of processed microbiome data

- "count-ish" data, minimum is zero
    + MetaPhlAn2 data are not actually integer counts
- non-normal
    + highly skewed (over-dispersed)
    + often has a lot of zero values
- samples differ in extraction and amplification efficiency, read depth
    + counts do not provide absolute microbial abundance
    + we can only infer _relative_ abundance

## Example: Rampelli Africa dataset

Rampelli S *et al.*: [Metagenome Sequencing of the Hadza Hunter-Gatherer Gut Microbiota](http://www.ncbi.nlm.nih.gov/pubmed/25981789). Curr. Biol. 2015, 25:1682–1693. 

```{r, message=FALSE, echo=FALSE}
suppressPackageStartupMessages(library(curatedMetagenomicData))
Rampelli = curatedMetagenomicData("RampelliS_2015.metaphlan_bugs_list.stool", 
                       dryrun = FALSE, counts = FALSE,
                       bugs.as.phyloseq = TRUE)[[1]]
```

```{r, echo=TRUE}
suppressPackageStartupMessages(library(phyloseq))
Rampelli
ab <- otu_table(Rampelli)
ab[1:5, 1:4]
summary(ab[, 1:4])
summary(sample_data(Rampelli))
```

# Distances in high-dimensional data analysis

## The importance of distance

- High-dimensional data are complex and impossible to visualize in raw form

<center>
<img src="animals.png" alt="animals" align="middle" style="height: 350px;">
</center>

## The importance of distance (cont'd)

- Distances can simplify thousands of dimensions
- Any clustering or classification of samples and/or features involves
combining or identifying objects that are close or similar.
- Distances or similarities are mathematical representations of what
we mean by close or similar.
- The choice of distance is a subject matter-specific, qualitative decision

## Heatmap

```{r Rampelliheatmap, warning=FALSE}
plot_heatmap(Rampelli, method="PCoA", distance="bray")
```

## Metrics and distances

A **metric** satisfies the following five properties:

1. non-negativity $d(a, b) \ge 0$
2. symmetry $d(a, b) = d(b, a)$
3. identification mark $d(a, a) = 0$
4. definiteness $d(a, b) = 0$ if and only if $a=b$
5. triangle inequality $d(a, b) + d(b, c) \ge d(a, c)$

- A **distance** is only required to satisfy 1-3.
- A **similarity function** satisfies 1-2, and **increases** as $a$ and $b$ become more similar
- A **dissimilarity function** satisfies 1-2, and **decreases** as $a$ and $b$ become more similar


## Euclidian distance (metric)

- Remember grade school:
```{r, echo=FALSE, fig.height=3.5}
rafalib::mypar()
plot(c(0,1,1),c(0,0,1),pch=16,cex=2,xaxt="n",yaxt="n",xlab="",ylab="",bty="n",xlim=c(-0.25,1.25),ylim=c(-0.25,1.25))
lines(c(0,1,1,0),c(0,0,1,0))
text(0,.2,expression(paste('(A'[x]*',A'[y]*')')),cex=1.5)
text(1,1.2,expression(paste('(B'[x]*',B'[y]*')')),cex=1.5)
text(-0.1,0,"A",cex=2)
text(1.1,1,"B",cex=2)
```
<center>
Euclidean d = $\sqrt{ (A_x-B_x)^2 + (A_y-B_y)^2}$.
</center>

- **Side note**: also referred to as *$L_2$ norm*

## Euclidian distance in high dimensions

```{r}
dim(otu_table(Rampelli))
table(sample_data(Rampelli)$location)
```

Interested in identifying similar *samples* and similar *microbes*

## Euclidian distance in high dimensions

- Points are no longer on the Cartesian plane,
- instead they are in higher dimensions. For example:
    - sample $i$ is defined by a point in 727 dimensional space: $(Y_{1,i},\dots,Y_{727,i})^\top$. 
    - feature $g$ is defined by a point in 38 dimensions $(Y_{g,38},\dots,Y_{g,38})^\top$

## Euclidian distance in high dimensions

Euclidean distance as for two dimensions. E.g., the distance between two samples $i$ and $j$ is:

$$ \mbox{dist}(i,j) = \sqrt{ \sum_{g=1}^{727} (Y_{g,i}-Y_{g,j })^2 } $$

and the distance between two features $h$ and $g$ is:

$$ \mbox{dist}(h,g) = \sqrt{ \sum_{i=1}^{38} (Y_{h,i}-Y_{g,i})^2 } $$


## 3 sample example

```{r}
ab <- otu_table(Rampelli)
dedauko1 <- ab[, 1]
dedauko2 <- ab[, 2]
bologna1 <- ab[, 29]
sqrt(sum((dedauko1 - dedauko2)^2))
sqrt(sum((dedauko1 - bologna1)^2))
```

## 3 sample example using dist()

```{r}
dim(ab)
(d <- dist(t(ab[, c(1, 2, 29)])))
class(d)
```

## Alpha / Beta diversity measures

From Morgan and Huttenhower [Human Microbiome Analysis](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002808)

<img src="journal.pcbi.1002808.g002.png" alt="SVD" align="middle" style="width: 1000px;">

These examples describe the A) sequence counts and B) relative abundances of six taxa detected in three samples. C) A collector's curve using a richness estimator approximates the relationship between the number of sequences drawn from each sample and the number of taxa expected to be present based on detected abundances. D) Alpha diversity captures both the organismal richness of a sample and the evenness of the organisms' abundance distribution. E) Beta diversity represents the similarity (or difference) in organismal composition between samples. 

* Shannon Index alpha diversity: $H' = -\sum_{i=1}^{S} \left( p_i ln(p_i) \right )$
* Beta diversity: $\beta = (n_1 - c) + (n_2 - c)$

## Beta diversity / dissimilarity 

E.g. Bray-Curtis dissimilarity between all pairs of samples:

```{r}
plot(hclust(phyloseq::distance(Rampelli, method="bray")), 
     main="Bray-Curtis Dissimilarity", xlab="", sub = "")
```

* Dozens of distance measures are available
    - see `?phyloseq::distance` and `?phyloseq::distanceMethodList`

# Dimension reduction and PCA

## Motivation for dimension reduction

Simulate the heights of twin pairs:
```{r}
suppressPackageStartupMessages(library(MASS))
set.seed(1)
n <- 100
y <- t(MASS::mvrnorm(n, c(0,0), matrix(c(1, 0.95, 0.95, 1), 2, 2)))
dim(y)
cor(t(y))
```

## Motivation for dimension reduction

```{r, echo=FALSE}
z1 = (y[1, ] + y[2, ]) / 2 #the sum 
z2 = (y[1, ] - y[2, ])   #the difference

z = rbind(z1, z2) #matrix now same dimensions as y

thelim <- c(-3,3)
rafalib::mypar(1,2)

plot(y[1, ], y[2, ], 
     xlab="Twin 1 (standardized height)",
     ylab="Twin 2 (standardized height)",xlim=thelim,
     ylim=thelim,
     main="Original twin heights")
points(y[1, 1:2], y[2, 1:2], col=2, pch=16)

plot(z[1, ], z[2, ], xlim=thelim, ylim=thelim,
     xlab="Average height", ylab="Difference in height",
     main="Manual PCA-like projection")
points(z[1, 1:2] , z[2, 1:2], col=2, pch=16)
```

## Motivation for dimension reduction

<center>
```{r, echo=FALSE, fig.height=3.5, fig.width=3.5}
rafalib::mypar()
d = dist(t(y))
d3 = dist(z[1,]) * sqrt(2) ##distance computed using just first dimension mypar(1,1)
plot(as.numeric(d), as.numeric(d3), xlab="Pairwise distances in 2 dimensions", 
     ylab="Pairwise distances in 1 dimension")
abline(0,1, col="red")
```
</center>

- Not much loss of height differences when just using average heights of twin pairs.
    - because twin heights are highly correlated

## Singular Value Decomposition (SVD)

SVD generalizes the example rotation we looked at:

$$\mathbf{Y} = \mathbf{UDV}^\top$$

<center>
<img src="SVD1.png" alt="SVD" align="middle" style="height: 350px;">
</center>

- **note**: the above formulation is for $m > n$

## Singular Value Decomposition (SVD)

<center>
<img src="SVD1.png" alt="SVD" align="middle" style="height: 275px;">
</center>

- $\mathbf{Y}$: the m rows x n cols matrix of measurements
- $\mathbf{U}$: m x n matrix relating original scores to PCA scores (**loadings**)
- $\mathbf{D}$: n x n diagonal matrix (**eigenvalues**)
- $\mathbf{V}$: n × n *orthogonal* matrix (**eigenvectors or PCA scores**)
    - orthogonal = unit length and "perpendicular" in 3-D

## SVD of Rampelli dataset

Scaling:
```{r, cache=TRUE}
e.standardize.fast <- t(scale(t(ab), scale=FALSE))
```

SVD:
```{r}
s <- svd(e.standardize.fast)
names(s)
```

## SVD of Rampelli dataset

```{r}
dim(s$u)     # loadings
length(s$d)  # eigenvalues
dim(s$v)     # d %*% vT = scores
```
<center>
<img src="SVD1.png" alt="SVD" align="middle" style="height: 200px;">
</center>

## SVD vs. prcomp()

```{r, cache=TRUE}
rafalib::mypar()
p <- prcomp( t(e.standardize.fast) )
plot(s$u[, 1] ~ p$rotation[, 1])
```

**Note:** u and v can each be multiplied by -1 arbitrarily

## PCA interpretation: loadings

<center>
<img src="SVD1.png" alt="SVD" align="middle" style="height: 225px;">
</center>

- $\mathbf{U}$ (**loadings**): relate the *principal component* axes to the original variables
    - think of principal component axes as a weighted combination of original axes

## PCA interpretation: loadings

```{r}
plot(p$rotation[, 1], xlab="Index of taxa", ylab="Loadings of PC1", 
     main="Loadings of PC1") #or, predict(p)
abline(h=c(-0.2, 0.2), col="red")
```

## PCA interpretation: loadings

- Taxa with high PC1 loadings
    + *note* look at taxanomic relationship of rows
    + should probably prune to one taxanomic level and re-do

```{r, fig.height=3.5}
e.pc1taxa <- ab[p$rotation[, 1] < -0.2 | p$rotation[, 1] > 0.2, ]
pheatmap::pheatmap(e.pc1taxa, scale="none", show_rownames=TRUE, 
                  show_colnames = FALSE)
```

## PCA interpretation: eigenvalues

- $\mathbf{D}$ (**eigenvalues**): standard deviation scaling factor that each decomposed variable is multiplied by.

```{r, fig.height=3, fig.width=5, echo=FALSE, fig.align='center'}
rafalib::mypar()
plot(p$sdev^2 / sum(p$sdev^2)*100, xlim=c(0, 40), type="b",
     ylab="% variance explained", main="Screeplot")
```

## PCA interpretation: eigenvalues

Alternatively as cumulative % variance explained (using `cumsum()` function):
```{r, fig.height=4, echo=FALSE, fig.align='center'}
rafalib::mypar()
plot(cumsum(p$sdev^2)/sum(p$sdev^2)*100, 
  ylab="cumulative % variance explained", ylim=c(0, 100), 
  type="b", main="Cumulative screeplot")
```

## PCA interpretation: scores

<center>
<img src="SVD1.png" alt="SVD" align="middle" style="height: 225px;">
</center>

- $\mathbf{V}$ (**scores**): The "datapoints" in the reduced prinipal component space
- In some implementations (like `prcomp()`), scores are $\mathbf{D V^T}$

## PCA interpretation: scores

```{r, fig.height=5, echo=FALSE}
loc = sample_data(Rampelli)$location
rafalib::mypar()
plot(p$x[, 1:2], xlab="PC1", ylab="PC2", 
     main="plot of p$x[, 1:2]",
     col=factor(loc), pch=as.integer(factor(loc)))
legend("bottomleft", legend=levels(factor(loc)), col=1:length(unique(loc)),
       pch=1:length(unique(loc)), bty='n')
```

## Principal Coordinates Analysis (PCoA)

- also referred to as Multi-dimensional Scaling (MDS) or *ordination*
- a reduced SVD, performed on a distance matrix
- identify two (or more) eigenvalues/vectors that preserve distances
- Advantage
    + PCA is limited to Euclidian distance, which is less useful for ecological data
    + compatible with any dissimilarity measure
- Disadvantage
    + no loadings to help with interpretation

## Principal Coordinates Analysis (cont'd)

```{r, echo=FALSE}
ord = ordinate(Rampelli, method="PCoA", distance="bray")
plot_ordination(Rampelli, ord, color="location", shape="location") + 
  ggplot2::ggtitle("Bray-Curtis Principal Coordinates Analysis")
```

* Available methods are "DCA", "CCA", "RDA", "CAP", "DPCoA", "NMDS", "MDS", "PCoA"

Not much "horseshoe" effect here.

# Within-sample diversity

## Alpha diversity estimates

* Look at `?phyloseq::estimate_richness`
* Supported measures of alpha diversity are:
    - "Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"
    - more information from `vegan` package

Note, you can ignore warning about singletons when performing this analysis.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
Rampelli.counts = curatedMetagenomicData("RampelliS_2015.metaphlan_bugs_list.stool", 
                       dryrun = FALSE, counts = TRUE,
                       bugs.as.phyloseq = TRUE)[[1]]
alpha_meas = c("Shannon", "Simpson", "InvSimpson")
(p <- plot_richness(Rampelli.counts, "gender", "location", measures=alpha_meas))
```

## Comparison of alpha diversity estimates

```{r, warning=FALSE, echo=FALSE}
alphas = estimate_richness(Rampelli.counts, measures=alpha_meas)
pairs(alphas)
```

## Links

- A built [html][] version of this lecture is available.
- The [source][] R Markdown is also available from Github.

[html]: http://rpubs.com/lwaldron/EPIC2017_introlecture
[source]: https://github.com/waldronlab/presentations/tree/master/Waldron_2017-06-30_EPIC
